### PR TITLE
[5.2] Add @set and @unset to Blade

### PR DESCRIPTION
**Pull request by sileence**
Since you added `@continue` and `@break` I think these ones could be useful as well. 

Adding variables to the view can be useful at a _presentation level_

For example in `welcome.blade.php` I want to add a custom title without using `@yield` / `@section` so I can do:

// welcome.blade.php
`@set ($title = "Welcome to website!")`

// layout.blade.php
And then `{{ $title or "Website" }}`

Or I could set a flag to show / hide some sections, add custom CSS classes, etc.

### CODE REVIEW

**Comment by arrilot** [comment]
@sileence #10503

File: src/Illuminate/View/Compilers/BladeCompiler.php [hunk 1]
```diff
@@ -707,6 +707,28 @@ protected function compileEndforelse($expression)
...
 
     /**
+     * Compile the set statements into valid PHP.
+     *
+     * @param  string  $expression
+     * @return string
+     */
+    protected function compileSet($expression)
+    {
+        return '<?php '.substr($expression, 1, -1).'; ?>';
```
Comment (GrahamCampbell) [inline src/Illuminate/View/Compilers/BladeCompiler.php]: Why is that needed?

Comment (sileence) [inline src/Illuminate/View/Compilers/BladeCompiler.php]: `($a = $b);` seemed weird to me #OCD

**Comment by sileence** [comment]
@arrilot never give up

File: src/Illuminate/View/Compilers/BladeCompiler.php (hunk 1 above)
Comment (GrahamCampbell) [inline src/Illuminate/View/Compilers/BladeCompiler.php]: Is it invalid code though?

Comment (sileence) [inline src/Illuminate/View/Compilers/BladeCompiler.php]: It's the same, so let me remove that then

**Comment by themsaid** [comment]
I just found out you can do this:

```
@section('title', 'Welcome to website!')
```

And later in the view:

```
@yield('title')
```

Not sure if it's clean enough, but it's doable.

**Comment by sileence** [comment]
@themsaid yet that won't work for IFs for example. 

Twig and Smarty have functions to set variables, I don't see why Blade shouldn't.

**Comment by themsaid** [comment]
Yeah you're right, it won't work inside php statements. You'll have to use:

```
$__env->yieldContent('title')
```

**Comment by arrilot** [comment]
> Twig and Smarty have functions to set variables, I don't see why Blade shouldn't.

Blade allows you to easily insert php code, while those engines don't

**Comment by sileence** [comment]
@arrilot inserting PHP directly doesn't look nice, and we could even create a macro to prevent that.

**Comment by boynet** [comment]
(I know that nobody asked me) but my opinion is to keep this framework to the minimum required stuff, and leave the rest to external plugins.. just like  with the HTML helper taking it out to external library was the good approach, the first reason most of us came to laravel it's because the ease of use and simplicity I think, if you can easily make a directive why you need it at the core level?

**Comment by sileence** [comment]
@boynet I understand that point, but it is not like that is going to make the framework bloated or anything like that. They are just 2 small directives.

**Comment by boynet** [comment]
@sileence yap but this 2 small directive and that 3 small stuff and now the docs are 30 lines longer

**Comment by sileence** [comment]
@boynet it doesn't mean you have to read the whole documentation or use them, they are just there in case you need them. I'm glad that Taylor decided to merge this.

**Comment by taylorotwell** [comment]
I renamed `set` to `php`:

```
@php($foo = 'bar');
```

Since that isn't really a "set" compile, it's just any raw PHP.

**Comment by sileence** [comment]
@taylorotwell I was thinking about that, they call that "do" in Twig, but "do" doesn't actually say anything.
